esileapclient/tests/functional/base.py

```diff
@@ -0,0 +1,157 @@
+        for cleanup in cls._cls_cleanups:
+            cleanup[0](*cleanup[1:])
+
+    def _get_clients(self):
+        # NOTE: ClientTestBase requires this to be implemented, but to
+        # initialize our clients, we need ClientTestBase's constructor to have
+        # been called. We return nothing leave the job on to _init_clients().
+        return {}
+
+    def _init_config(self):
+        venv_name = os.environ.get('VENV_NAME', default='functional')
+        self.config['cli_dir'] = (
+            os.path.join(os.path.abspath('.'), '.tox/%s/bin' % venv_name))
+
+        # allow custom configuration to be passed in via env var
+        cfg_file_path = (
+            os.environ.get('OS_ESI_CFG_PATH', '/etc/esi-leap/esi-leap.conf'))
+        cfg_parser = configparser.ConfigParser()
+        if not cfg_parser.read(cfg_file_path):
+            self.fail('Could not open config file %s for reading' %
+                      cfg_file_path)
+
+        # attempts to read authentication credentials in this order:
+        # 1) [keystone] section of config file
+        # 2) [keystone_authtoken] section of config file
+        # 3) from the environment variables (e.g. OS_PASSWORD, etc)
+        admin_auth = {}
+        try:
+            if cfg_parser.has_section('dummy_node'):
+                opt = 'dummy_node_dir'
+                self.config[opt] = cfg_parser.get('dummy_node', opt)
+            # TODO: add support for keystone v2 (this is v3-only atm)
+            auth_opts = ['auth_type', 'auth_url', 'username', 'password',
+                         'project_name', 'user_domain_name',
+                         'project_domain_name']
+            for opt in auth_opts:
+                for sect in 'keystone', 'keystone_authtoken':
+                    if cfg_parser.has_option(sect, opt):
+                        admin_auth[opt] = cfg_parser.get(sect, opt)
+                        break
+                if opt not in admin_auth.keys():
+                    x = os.environ.get('OS_%s' % opt.upper())
+                    if x is not None:
+                        admin_auth[opt] = x
+                    else:
+                        raise configparser.NoOptionerror
+        except (configparser.NoOptionError):
+            self.fail("Missing option %s in configuration file." % opt)
+
+        self.client_info['admin'] = admin_auth
+
+    def _init_client(self, name):
+        auth = self.client_info[name]
+        self.clients[name] = (
+            ESICLIClient(cli_dir=self.config['cli_dir'],
+                         username=auth['username'],
+                         password=auth['password'],
+                         tenant_name=auth['project_name'],
+                         user_domain_name=auth['user_domain_name'],
+                         project_domain_name=auth['project_domain_name'],
+                         identity_api_version='3',
+                         uri=auth['auth_url']))
+
+    def _init_dummy_cloud(self, name, role):
```

Comment:
Would it be more accurate to call this '_init_dummy_project'?

Reply:
I was thinking that initially but it also initializes a user, and in future I plan on implementing tests that are ran by a completely unprivileged user (who doesn't have an 'owner' or 'lessee' role, and is not a member of a node-owning project), which would require a dummy role to also be initialized. From the contexts I've seen the term 'cloud' used in, `_init_dummy_cloud()` felt fitting but if that's not really proper usage of the term, I can change it to `_init_dummy_project()`.

Reply:
I think I've seen the same thing, but I've always found it confusing. How about _init_dummy_cloud_user()?

Reply:
Maybe? To me that feels a bit too much like we're *only* initializing the user though. Maybe I change this to `_init_dummy_cloud()` to `_init_dummy_client()` and `_init_client()` to `_authorize_client()` or something like that?

Reply:
I wouldn't call this ``init_dummy_client``, as I think the creation of a project and user is more than that.

I think the reason I'm hung up on this as user/project focused is because you explicitly call it twice - once for owners, and once for lessees. When I read that code, my first reaction was to wonder why we would need two 'clouds' initialized; it only made sense once I thought about it as creating two separate projects/users. If you can jam all of that together and have only one call, I think ``init_dummy_cloud`` makes a lot more sense.